feat(contributing): add task scheduling (#282)
autowarefoundation/autoware-documentation: docs/contributing/coding-guidelines/ros-nodes/task-scheduling.md

# Task scheduling

## Scheduing in autoware

> The software of autoware is the system of multiple nodes.In ROS2 the executor uses one or more threads of the underlying operating system to invoke the callbacks. The different type of callback has the different priority.  
> In autoware,there are two types about "publish-subscribe":

Reviewer: ```suggestion
Autoware is a system made of multiple ROS2 nodes.
In ROS2 the executor uses one or more threads of the underlying operating system to invoke the callbacks.
Callbacks are units of work like subscription callbacks, timer callbacks, service calls, and received client responses.
The different types of callback has the different priority. Here are the priorities for:
```
Reference documentation: https://docs.ros2.org/latest/api/rclpy/api/execution_and_callbacks.html#module-rclpy.executors

Also made some small fixes like spaces after commas and periods.

    node_ptr->create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
```

The type of MutuallyExclusive ensure that callbacks of this group must not be executed in parallel.You can see more information about scheduling of ROS2 in theses links:

- <https://docs.ros.org/en/humble/Concepts/About-Executors.html>
- <https://docs.ros.org/en/humble/How-To-Guides/Using-callback-groups.html>

## Analyse of data flow and running time

In autoware,there are many topics used by different nodes.You can analyse the data flow and running time of multiple nodes in the whole software. By the analyse,you can intuitively understand the nodes operation path and input/output flow.You can use CARET or TILDE of autoware to finish the analyse.

### CARET

By using caret,you can trace the application without changing codes of application.You can analyse the latency and the running time of nodes.  
You can install TIDLE by following steps:

1. Clone caret and enter the directory.

   ```bash
       git clone https://github.com/tier4/caret.git ros2_caret_ws
       cd ros2_caret_ws
   ```

2. Create the src directory and clone repositories into it.

   ```bash
       mkdir src

>        vcs import src < caret.repos --recursive

Reviewer 2: `--recursive` is not necessary.

```suggestion
       vcs import src < caret.repos
```

   ```

3. Run setup_caret.sh.

   ```bash
       ./setup_caret.sh
   ```

4. Build the workspace.

   ```bash

>        source /opt/ros/galactic/setup.bash(or source /opt/ros/humble/setup.bash)

Reviewer 2: Installation of caret for galactic should not be mentioned here, I think.

```suggestion
       source /opt/ros/humble/setup.bash
```
